### PR TITLE
netdev/upper: Optimize on quota related operations

### PR DESCRIPTION
**Pull request by wengzhe**
## Summary
Patches included:
- netdev/upper: Make quota_load public
  - In case some lower-half need to get current quota, like https://github.com/apache/nuttx/pull/9297
- netdev/upper: Use atomic when we have atomic support from compiler
  - Checks whether current compiler supports atomic, and use atomic instead of spin_lock to optimize performance.

## Impact
netdev upper half, currently used on sim only

## Testing
sim and https://github.com/apache/nuttx/pull/9297

### CODE REVIEW

**Review by acassis — APPROVED** [review]

File: drivers/net/netdev_upperhalf.c [hunk 1]
```diff
@@ -72,6 +84,10 @@ struct netdev_upperhalf_s
 #else
   struct work_s work;
 #endif
+
+#ifdef HAVE_ATOMIC
+  atomic_int quota[NETPKT_TYPENUM]; /* A copy as atomic from lower-half */
```
Comment (xiaoxiang781216) [inline drivers/net/netdev_upperhalf.c:89]: why not change netdev_lowerhalf_s::quota to atomic_int directly in this case?

Comment (wengzhe) [reply drivers/net/netdev_upperhalf.c:89]: Then the lower-half will need to change their init expression to `atomic_init`. Or maybe we can make a macro for lower-half to init it?

Comment (xiaoxiang781216) [reply drivers/net/netdev_upperhalf.c:89]: is it fine to assign the atomic variable directly in the initial phase?

Comment (wengzhe) [reply drivers/net/netdev_upperhalf.c]: Yes, it seems fine, try to do like this.

File: include/nuttx/net/netdev_lowerhalf.h [hunk 2]
```diff
@@ -234,6 +258,21 @@ void netdev_lower_rxready(FAR struct netdev_lowerhalf_s *dev);
...
+ * Description:
+ *   Fetch the quota, use this interface when device is running.
+ *
+ * Input Parameters:
+ *   dev  - The lower half device driver structure
+ *   type - Whether get quota for TX or RX
+ *
+ ****************************************************************************/
+
+int netdev_lower_quota_load(FAR struct netdev_lowerhalf_s *dev,
```
Comment (pkarashchenko) [inline include/nuttx/net/netdev_lowerhalf.h:273]: What is the reason for exposing this interface? Is it going to be used by third party code in kernel?

**Comment by pkarashchenko** [comment]
@wengzhe could you please try https://github.com/apache/nuttx/pull/9552 ?
It seems to be natural to me, but I do not have an ability to test it.

File: include/nuttx/net/netdev_lowerhalf.h (hunk 2 above)
Comment (xiaoxiang781216) [reply include/nuttx/net/netdev_lowerhalf.h:273]: Yes, it's used here: https://github.com/apache/nuttx/pull/9297/files#diff-fa7cc8834e3748b89579efd930a809981c7267ef5aa0f573916ebe57ed1a2f71R319

**Pull request merged**